### PR TITLE
Storybook test integration

### PR DESCRIPTION
**Pull request by Gkrumbach07**
fixes: https://github.com/opendatahub-io/odh-dashboard/issues/962

Configured storybooks for testing in the future. I added MSW add on, accessibly add on, and test-runner addon.

## Description

You can now generate a local dev storybook by running `npm run storybook` in the frontend folder. You can navigate through each component and select stories. A story in our case will represent a user flow we want to test.

For now the example is looking at the entire project page defined in `ProjectView.tsx`. The stories in there include case for creating, editing, and deleting. 

Each story has interaction tests and a11y accessibility tests.

You can run the interaction tests in the terminal with `npm run test:storybook` inside the frontend folder. The backend is not connected for storybooks as it is only looking at frontend tests. So there are mocks created for most of the `GET` requests.

image
image

## How Has This Been Tested?

To test run `npm run test:storybook` in the frontend folder. All tests should pass.
To see storybook run `npm run storybook` in frontend folder. data should populate. (note going to other routes will not work)

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by Gkrumbach07** [comment]
I have kept the old tests functioning, however we may want to move them over to storybooks if they are rendering in any way

File: frontend/.storybook/preview.js [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
...ablement\":true},\"groupsConfig\":{\"adminGroups\":\"rhods-admins\",\"allowedGroups\":\"system:authenticated\"},\"modelServerSizes\":[{\"name\":\"Small\",\"resources\":{\"limits\":{\"cpu\":\"2\",\"memory\":\"8Gi\"},\"requests\":{\"cpu\":\"1\",\"memory\":\"4Gi\"}}},{\"name\":\"Medium\",\"resources\":{\"limits\":{\"cpu\":\"8\",\"memory\":\"10Gi\"},\"requests\":{\"cpu\":\"4\",\"memory\":\"8Gi\"}}},{\"name\":\"Large\",\"resources\":{\"limits\":{\"cpu\":\"10\",\"memory\":\"20Gi\"},\"requests\":{\"cpu\":\"6\",\"memory\":\"16Gi\"}}}],\"notebookController\":{\"enabled\":true,\"notebookNamespace\":\"rhods-notebooks\",\"pvcSize\":\"20Gi\"},\"notebookSizes\":[{\"name\":\"Small\",\"resources\":{\"limits\":{\"cpu\":\"2\",\"memory\":\"8Gi\"},\"requests\":{\"cpu\":\"1\",\"memory\":\"8Gi\"}}},{\"name\":\"Medium\",\"resources\":{\"limits\":{\"cpu\":\"6\",\"memory\":\"24Gi\"},\"requests\":{\"cpu\":\"3\",\"memory\":\"24Gi\"}}},{\"name\":\"Large\",\"resources\":{\"limits\":{\"cpu\":\"14\",\"memory\":\"56Gi\"},\"requests\":{\"cpu\":\"7\",\"memory\":\"56Gi\"}}},{\"name\":\"X Large\",\"resources\":{\"limits\":{\"cpu\":\"30\",\"memory\":\"120Gi\"},\"requests\":{\"cpu\":\"15\",\"memory\":\"120Gi\"}}}]}}\n"
```
Comment (andrewballantyne) [inline frontend/.storybook/preview.js:39]: Can we not trim this to readable values instead of jamming in the k8s nonsense?

Ie, delete `kubectl.kubernetes.io/last-applied-configuration` and `managedFields`

File: frontend/.storybook/preview.js [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+                }
+              }
+            ]
+          }
+        },
+      }}
+    >
+      <Provider store={store}>
+        <Router>
+          <SDKInitialize>
```
Comment (andrewballantyne) [inline frontend/.storybook/preview.js:254]: Do we not have to concern ourselves with this not being mocked?

File: frontend/__mocks__/mockNotebooks.ts [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+                },
+                "creationTimestamp": "2023-02-14T21:44:13Z",
+                "generation": 4,
+                "labels": {
+                    "app": "workbench",
+                    "opendatahub.io/dashboard": "true",
+                    "opendatahub.io/odh-managed": "true",
+                    "opendatahub.io/user": "admin"
+                },
+                "managedFields": [
```
Comment (andrewballantyne) [inline frontend/__mocks__/mockNotebooks.ts:28]: I'd really like to see this not in our mocks -- is this generated?

(this applies to all mocks)

File: frontend/__mocks__/mockWorkbench.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
+// /api/k8s/apis/route.openshift.io/v1/namespaces/project/routes/workbench
+
+export const mockWorkbench = {
```
Comment (andrewballantyne) [inline frontend/__mocks__/mockWorkbench.ts:3]: mock ... Route?

File: frontend/config/mockServiceWorker.js [hunk 5]
```diff
@@ -0,0 +1,303 @@
+/* eslint-disable */
+/* tslint:disable */
+
+/**
+ * Mock Service Worker (1.0.1).
```
Comment (andrewballantyne) [inline frontend/config/mockServiceWorker.js:5]: Do we mock service workers for storybook? As far as I know we don't use them for the app.

File: frontend/config/setupTests.js [hunk 6]
```diff
@@ -17,3 +19,7 @@ configure({
 });
 
 jest.mock('@openshift/dynamic-plugin-sdk-utils', () => ({}));
+
+
+
+global.TextEncoder = TextEncoder;
```
Comment (andrewballantyne) [inline frontend/config/setupTests.js:25]: Can we make sure we have newlines at the end of files? I bet our linter isn't catching this because we are not looking at JS files... might be worth looking into.

File: frontend/package.json [hunk 7]
```diff
@@ -66,11 +70,23 @@
     "yaml": "^2.1.3"
   },
   "devDependencies": {
+    "@babel/core": "^7.21.0",
+    "@storybook/addon-a11y": "^6.5.16",
```
Comment (andrewballantyne) [inline frontend/package.json:74]: Oooooh, nice! How much is this going to help us?

File: frontend/package.json [hunk 8]
```diff
@@ -33,7 +33,10 @@
     "test:fix": "eslint --ext .json,.js,.ts,.jsx,.tsx ./src --fix",
     "test:lint": "eslint --max-warnings 0 --ext .json,.js,.ts,.jsx,.tsx ./src",
     "test:local": "jest --watchAll",
-    "test:type-check": "tsc --noEmit"
+    "test:type-check": "tsc --noEmit",
+    "test:storybook": "test-storybook --verbose",
+    "storybook": "start-storybook -p 6006 -s ./config",
```
Comment (andrewballantyne) [inline frontend/package.json:38]: Can we get this added to the root README? I'd imagine it can go right after `npm test`

File: frontend/package.json [hunk 9]
```diff
@@ -33,7 +33,10 @@
     "test:fix": "eslint --ext .json,.js,.ts,.jsx,.tsx ./src --fix",
     "test:lint": "eslint --max-warnings 0 --ext .json,.js,.ts,.jsx,.tsx ./src",
     "test:local": "jest --watchAll",
-    "test:type-check": "tsc --noEmit"
+    "test:type-check": "tsc --noEmit",
+    "test:storybook": "test-storybook --verbose",
```
Comment (andrewballantyne) [inline frontend/package.json:37]: Interesting -- this doesn't seem to fail on a11y issues -- shouldn't it?

File: frontend/package.json [hunk 10]
```diff
@@ -33,7 +33,10 @@
...
+    "storybook": "start-storybook -p 6006 -s ./config",
+    "build-storybook": "build-storybook"
```
Comment (andrewballantyne) [inline frontend/package.json:39]: What would be the use-case for building it?

File: frontend/src/pages/projects/screens/projects/ProjectView.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  expect(deleteButton).not.toBeDisabled();
+};
+
+export const CreateProject = Template.bind({});
+CreateProject.parameters = {
+  a11y: {
+    element: '.pf-c-backdrop',
+  },
+};
+CreateProject.play = async ({ canvasElement }) => {
```
Comment (andrewballantyne) [inline frontend/src/pages/projects/screens/projects/ProjectView.stories.tsx:128]: Interesting -- going to need a crash course on what's going on here. Got links to docs to explain what is happening?

Since this is a common pattern, did we want a utility to setup a test?

---

`Template.bind` feels ... old. This is just a hack to duplicating an instance so you can mutate it?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Looking awesome! Thanks for doing this up Gage! I have some comments and questions.

Also, is there anyway we can get the test not to require me to start an instance? Just for ease of use -- could it start one and then use it?

File: frontend/package.json (hunk 10 above)
Comment (Gkrumbach07) [reply frontend/package.json:39]: we would build it if we wanted to deploy a hosted version of storybook. (could be a form of docs)

File: frontend/__mocks__/mockWorkbench.ts (hunk 4 above)
Comment (Gkrumbach07) [reply frontend/__mocks__/mockWorkbench.ts:3]: its a mock for a route GET 200 response. It uses [mock service worker](https://mswjs.io/) to apply the mocks to the network calls. I found it makes it easier to control what is going on without having to put mocks on every function that calls the api

File: frontend/config/mockServiceWorker.js (hunk 5 above)
Comment (Gkrumbach07) [reply frontend/config/mockServiceWorker.js:5]: yes, this is just used for storybook. I dont think it will run the mocks in dev/prod, b/c the service worker is only initiated in the .storybook/preview.js file

File: frontend/__mocks__/mockNotebooks.ts (hunk 3 above)
Comment (andrewballantyne) [reply frontend/__mocks__/mockNotebooks.ts:28]: `managedFields` and any k8s-configuration annotation -- effectively the large blocks of k8s-esk things. They have no value but to bloat the code and make it harder to read the real values / update the values.

File: frontend/package.json (hunk 7 above)
Comment (Gkrumbach07) [reply frontend/package.json:74]: A ton. It essentially runs a11y tests on any "story" we create. So as logn as we contunie to create stories for new (or old) components, this test can be run. 

we can use it in parralell with dev by having the storybook run in watch mode.
we can add it to the storybook testing cli, so both interaction and accessibility tests run
we can have it then integrate with CI

File: frontend/__mocks__/mockWorkbench.ts (hunk 4 above)
Comment (andrewballantyne) [reply frontend/__mocks__/mockWorkbench.ts:3]: You called it `mockWorkbench` but it returns a Route object

File: frontend/package.json (hunk 8 above)
Comment (Gkrumbach07) [reply frontend/package.json:38]: i didnt right away just so we could slowly adopt this

File: frontend/package.json (hunk 9 above)
Comment (Gkrumbach07) [reply frontend/package.json:37]: right now no. I can add that though. I didnt know if this was the way we wanted to test a11y yet

Comment (andrewballantyne) [reply frontend/package.json:37]: Anyway to get additional testing blocks sounds like a plan to me.

File: frontend/package.json (hunk 8 above)
Comment (andrewballantyne) [reply frontend/package.json:38]: Might as well add it with this PR -- shouldn't be too much documentation needed to just note it exists.

File: frontend/.storybook/preview.js (hunk 1 above)
Comment (Gkrumbach07) [reply frontend/.storybook/preview.js:39]: Yes we probably can. This is jsut the raw api response.

File: frontend/__mocks__/mockNotebooks.ts (hunk 3 above)
Comment (Gkrumbach07) [reply frontend/__mocks__/mockNotebooks.ts:28]: I will go through the mocks and trim them

File: frontend/__mocks__/mockWorkbench.ts (hunk 4 above)
Comment (Gkrumbach07) [reply frontend/__mocks__/mockWorkbench.ts:3]: that is whatever was returned by that api route commented above the file.

File: frontend/package.json (hunk 9 above)
Comment (Gkrumbach07) [reply frontend/package.json:37]: I will add an update commit with that implementation

File: frontend/src/pages/projects/screens/projects/ProjectView.stories.tsx (hunk 11 above)
Comment (Gkrumbach07) [reply frontend/src/pages/projects/screens/projects/ProjectView.stories.tsx:128]: https://storybook.js.org/docs/6.5/react/writing-stories/args/

This talks about how a story is constructed. 

TLDR
The template is reused for each story.

File: frontend/__mocks__/mockWorkbench.ts (hunk 4 above)
Comment (Gkrumbach07) [reply frontend/__mocks__/mockWorkbench.ts:3]: it is no. that is just a name i gave it. I can do something like `mock<resource_type><name>`

**Comment by Gkrumbach07** [comment]
>Also, is there anyway we can get the test not to require me to start an instance? Just for ease of use -- could it start one and then use it?

There is but the solution required 3 more dev deps (to run the command concurrently and wait for it to load, then tear it down). So i did not add that feature, but can if those deps are fine with you

**Comment by andrewballantyne** [comment]
> > Also, is there anyway we can get the test not to require me to start an instance? Just for ease of use -- could it start one and then use it?
> 
> There is but the solution required 3 more dev deps (to run the command concurrently and wait for it to load, then tear it down). So i did not add that feature, but can if those deps are fine with you

@Gkrumbach07 nah -- lets follow up on it. No worries!

File: frontend/src/__mocks__/mockAPINamespaceProjects.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
+export const mockAPINamespaceProjects = {
```
Comment (andrewballantyne) [inline frontend/src/__mocks__/mockAPINamespaceProjects.ts:1]: Shouldn't there be a comment at the top of this file for the api path?

File: frontend/src/__mocks__/mockDashboardConfig.ts [hunk 13]
```diff
@@ -0,0 +1,140 @@
+export const mockDashboardConfig = {
```
Comment (andrewballantyne) [inline frontend/src/__mocks__/mockDashboardConfig.ts:1]: Same here...

**Review by andrewballantyne — COMMENTED** [review]
Nits for consistency... let me test this PR out again.
